Common/include/parallelization/vectorization.hpp

```diff
@@ -0,0 +1,212 @@
+/*!
+ * \file vectorization.hpp
+ * \brief Implementation of a portable SIMD type.
+ * \author P. Gomes
+ * \version 7.0.5 "Blackbird"
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2020, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "../basic_types/datatype_structure.hpp"
+#include "../omp_structure.hpp"
+#include <initializer_list>
+#include <cmath>
+#include <algorithm>
+
+namespace simd {
+
+/*--- Detect preferred SIMD size (bytes). ---*/
+#if defined(__AVX512F__)
+constexpr size_t SIMD_SIZE = 64;
+#elif defined(__AVX__)
+constexpr size_t SIMD_SIZE = 32;
+#elif defined(__SSE__)
+constexpr size_t SIMD_SIZE = 16;
+#else
+constexpr size_t SIMD_SIZE = 8;
+#endif
+
+/*!
+ * \brief Convert the SIMD size (bytes) to a lenght (num elems).
+ */
+template<class T>
+constexpr size_t simdLen() { return SIMD_SIZE / sizeof(T); }
+template<>
+constexpr size_t simdLen<su2double>() { return SIMD_SIZE / sizeof(passivedouble); }
+
+/*!
+ * \class Array
+ * \brief A simple SIMD type relying mostly on implicit vectorization, i.e. done by
+ * the compiler, as opposed to explicit (done via intrinsics or inline assembly).
+ */
+template<class Scalar_t, size_t N = simdLen<Scalar_t>()>
+class Array {
```

Comment:
The vectorization will be based on a "simd type" basically a static array class for which we overload all required operators and functions.
For now this relies just on the compiler to do "the right thing" which does require a few more flags than folks will be used to, for example https://gcc.godbolt.org/z/N4FnXM.
Relying just on the compiler makes it easier to extend functionality and makes the code compatible with un-vectorizable types.
The important thing is to hide the implementation behind a generic interface, then we can think about specializing key operations on certain types using architecture-specific intrinsics for example.